### PR TITLE
impl call base

### PR DESCRIPTION
**Pull request by BusyJay**
Add basic call functions, client call and server call will be implemented in other prs respectively.

### CODE REVIEW

File: src/async/promise.rs [hunk 1]
```diff
@@ -14,9 +14,90 @@
...
+
+    fn finish_response(&mut self, succeed: bool) {
+        let mut guard = self.inner.lock();
+        if !succeed {
+            guard.set_result(Err(Error::RemoteStopped));
+            return;
+        }
+        let status = self.ctx.rpc_status();
+        if status.status != GrpcStatusCode::Ok {
+            guard.set_result(Err(Error::RpcFailure(status)));
```
Comment (siddontang) [inline src/async/promise.rs]: only status, not other error messages here?

Comment (BusyJay) [inline src/async/promise.rs]: See https://github.com/pingcap/grpc-rs/pull/15/files#diff-8c8308f2c854f39735f331e70f94d8d2R51

File: src/call/mod.rs [hunk 2]
```diff
@@ -34,3 +64,296 @@ impl RpcStatus {
...
+    }
+}
+
+impl Drop for BatchContext {
+    fn drop(&mut self) {
+        unsafe { grpc_sys::grpcwrap_batch_context_destroy(self.ctx) }
+    }
+}
+
+/// A helper function theat run the batch call and check the result.
```
Comment (siddontang) [inline src/call/mod.rs]: /theat/that/s

Comment (siddontang) [inline src/call/mod.rs]: runs and checks

File: src/async/promise.rs [hunk 3]
```diff
@@ -14,9 +14,90 @@
...
+            return;
+        }
+
+        guard.set_result(Ok(self.ctx.recv_message()))
+    }
+
+    pub fn resolve(mut self, success: bool) {
+        match self.ty {
+            BatchType::CheckRead => {
+                assert!(success);
```
Comment (siddontang) [inline src/async/promise.rs]: always true here?

File: src/async/promise.rs [hunk 4]
```diff
@@ -14,9 +14,90 @@
...
+        }
+    }
+
+    pub fn context(&self) -> &BatchContext {
+        &self.ctx
+    }
 
+    fn read_one_msg(&mut self) {
+        let mut guard = self.inner.lock();
+        guard.set_result(Ok(self.ctx.recv_message()));
```
Comment (siddontang) [inline src/async/promise.rs]: can recv_message cost a long time?

File: src/call/mod.rs [hunk 5]
```diff
@@ -34,3 +64,296 @@ impl RpcStatus {
...
+                                                            send_empty_metadata,
+                                                            payload_ptr as _,
+                                                            payload_len,
+                                                            write_flags,
+                                                            tag)
+        })
+    }
+
+    /// Cancel the rpc call by client.
+    fn cancel(&self) {
```
Comment (siddontang) [inline src/call/mod.rs]: where to use it?

File: src/call/mod.rs [hunk 6]
```diff
@@ -34,3 +64,296 @@ impl RpcStatus {
...
+            try_ready!(write_f.poll());
+        }
+
+        self.write_f.take();
+        Ok(Async::Ready(()))
+    }
+
+    fn close(&mut self, call: &mut Call) -> Poll<(), Error> {
+        if self.close_f.is_none() {
+            if let Async::NotReady = try!(self.poll_complete()) {
```
Comment (siddontang) [inline src/call/mod.rs]: can we use `try_ready!` here?

File: src/async/promise.rs (hunk 3 above)
Comment (BusyJay) [inline src/async/promise.rs]: Yes.

File: src/async/promise.rs (hunk 4 above)
Comment (BusyJay) [inline src/async/promise.rs]: No, just copy serveral pieces of memories together.

File: src/call/mod.rs (hunk 5 above)
Comment (BusyJay) [inline src/call/mod.rs]: When wants to cancel the rpc call.

Comment (siddontang) [inline src/call/mod.rs]: should be `pub`?

Comment (BusyJay) [inline src/call/mod.rs]: No, it won't be called by user directly.

**Comment by siddontang** [comment]
Any test or add later?

**Comment by BusyJay** [comment]
When call is used, completion queue is always involved. And relay on the call order. So it's better add integration tests later.

File: src/call/mod.rs (hunk 6 above)
Comment (siddontang) [inline src/call/mod.rs]: can we cancel the write directly, no need to wait to write completed?

**Comment by siddontang** [comment]
PTAL @disksing @andelf

File: src/call/mod.rs (hunk 6 above)
Comment (BusyJay) [inline src/call/mod.rs]: No. On the one hand, future requires buffered data should be flushed before actually closing:

> if a sink has buffered data, then it'll be flushed out during a close operation.

On the other hand, not flushing data may lead to wrong result for the streaming call.

File: src/call/mod.rs [hunk 7]
```diff
@@ -12,7 +12,37 @@
...
+
+use async::{BatchFuture, BatchType, Promise};
+use error::{Error, Result};
+
+#[derive(Clone, Copy)]
+pub enum MethodType {
+    Unary,
+    ClientStreaming,
+    ServerStreaming,
+    Dulex,
```
Comment (disksing) [inline src/call/mod.rs]: Duplex, not Dulex.

Comment (siddontang) [inline src/call/mod.rs]: Ping @BusyJay

**Comment by andelf** [comment]
LGTM. 
Promises as traits may be better, enum may be misleading if you want to add more variants.

**Comment by BusyJay** [comment]
As [comment](https://github.com/pingcap/grpc-rs/pull/15/files#diff-2112f55f2a2d7b22cfa3be55aaa9e47cR108) says, it's not suitable to use generic or trait here.

**Comment by siddontang** [comment]
PTAL @disksing

File: src/call/mod.rs [hunk 8]
```diff
@@ -34,3 +64,307 @@ impl RpcStatus {
...
+    where F: FnOnce(*mut GrpcBatchContext, *mut c_void) -> GrpcCallStatus
+{
+    let (cq_f, prom) = Promise::batch_pair(bt);
+    let prom_box = Box::new(prom);
+    let batch_ptr = prom_box.batch_ctx().unwrap().as_ptr();
+    let prom_ptr = Box::into_raw(prom_box);
+    let code = f(batch_ptr, prom_ptr as *mut c_void);
+    if code != GrpcCallStatus::Ok {
+        unsafe {
+            Box::from_raw(prom_ptr);
```
Comment (overvenus) [inline src/call/mod.rs]: For freeing the `prom_ptr`?

File: src/call/mod.rs [hunk 9]
```diff
@@ -34,3 +64,307 @@ impl RpcStatus {
...
+        }
+        panic!("create call fail: {:?}", code);
```
Comment (overvenus) [inline src/call/mod.rs]: Why not return errors?

File: src/call/mod.rs (hunk 8 above)
Comment (BusyJay) [inline src/call/mod.rs]: Yes.

File: src/call/mod.rs (hunk 9 above)
Comment (BusyJay) [inline src/call/mod.rs]: Because it won't return error code if implemented correctly.

**Comment by siddontang** [comment]
PTAL @disksing @overvenus

**Pull request merged**